### PR TITLE
Use org.postgresql jdbc driver

### LINKED ISSUE #819: Evaluate postgres jdbc drivers
The output of this ticket should be a trade study, with pros and cons, and a recommendation for which driver we should be using.

The trade study and decision should be summarized in the Aerie SDD.

Currently we use https://impossibl.github.io/pgjdbc-ng/ which claims to be the fastest postgres jdbc driver.

The "official" postgres jdbc driver seems to have been recently revamped, and may be a more attractive option now: https://jdbc.postgresql.org/

### PR DESCRIPTION
**Pull request by mattdailis**
* **Tickets addressed:** Related to #819 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

For this PR, I replaced impossibl.pgjdbc-ng with org.postgresql, and made the updates necessary to get the tests to pass.

Most changes were related to type casting in SQL (apparently pgjdbc-ng did some magic with inferring types), and replacing uses of pgjdbc-ng's Interval with PGInterval.

The listener had to be implemented slightly differently - though it ended up being fairly straightforward.

## Verification

Mostly relying on e2e-tests - I also did some manual spot checks. Would be good to get some more comprehensive manual testing done before merging.

- [x] More comprehensive manual testing

## Documentation

If we choose to pursue this route, I'll update the SDD:

- [ ] Update SDD

## Future work

None

### CODE REVIEW

**Comment by Twisol** [comment]
I wonder if it's possible to get this battle-tested a bit before we merge? It would be nice to have some confidence that (1) the issue we saw previously isn't happening, (2) there aren't any obvious strange new bugs in the new driver, and (3) things still work pretty much as normal.

(I wonder if there are any performance implications for storing simulation results, in particular.)

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java
Comment (Twisol) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java]: Oh, this is very depressing. Do you know what specific error comes up if you leave the type annotation off?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:193]: Why the strong preference for `res.next` over `res.first`?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Why was `snapshotIds` changes from an int[] to an ArrayList?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: This change removes the assertion described in the comment. The old code checked that there was exactly one row. The new code checks that the first row's first column equals 1.

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: If this is a combination of `//assert that the snapshot history is n+1 long` and `//assert that res contains, in order: finalSnapshotId, snapshotId[0,1,...,n]`, then the comment should reflect that.

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java
Comment (Mythicaeda) [inline scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java]: This `break` is only going to break out of the `for` loop. If this thread is interrupted, we likely want to break out of the `while(true)` loop above (meaning the condition would be `while(!Thread.interrupted())`. Otherwise, the thread won't respond to its `interrupted()` flag being set on the line above.

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java
Comment (Mythicaeda) [inline merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java:44]: Interesting thing I noticed when looking at the [Hikari docs](https://www.javadoc.io/doc/com.zaxxer/HikariCP/latest/com/zaxxer/hikari/HikariDataSource.html): `HikariDataSource` has a `close` method that we never call. Although given that the while loop below only breaks on SIGINT or SIGKILL, I'm not sure when we _could_ call it 🤔.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java:1]: This class should be deleted, as it's unused as of #723.

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/graphql/GraphQLParsers.java
Comment (Mythicaeda) [inline scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/graphql/GraphQLParsers.java:1]: The an `intervalParser` Pattern defined near the top of this class that now goes unused.

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java
Comment (mattdailis) [reply scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java]: Good catch, thank you

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:193]: Good question - according to the JDBC spec, by default ResultSets are "forward only", meaning that they support only the `next` operation. 

https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html

Here's the place in the new driver that sets that default:

https://github.com/pgjdbc/pgjdbc/blob/3c3e5b62349c42d9d3dfe1fe0102776d4de8c353/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L515-L519

And here's where it throws an exception if you try to scroll a non-scrollable result set:

https://github.com/pgjdbc/pgjdbc/blob/3c3e5b62349c42d9d3dfe1fe0102776d4de8c353/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L312-L319

So, the fact that impossibl's pgjdbc-ng allows us to call `first`, even though the statement's type is "forward only", is a deviation from the spec.

If we want to be able to call `first`, we can specify that when creating the statement, it would look something like this:

```java
connection.createStatement(
        ResultSet.TYPE_SCROLL_SENSITIVE,
        ResultSet.CONCUR_READ_ONLY);
```

My approach in this PR was to update our code to work with the default non-scrollable result sets

Comment (Twisol) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:193]: ooooh, good find

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: My intent here was to be able to directly compare the two lists for equality - in the event that the assertion fails, it provides both lists in the message, rather than just the element that was unequal.

https://github.com/NASA-AMMOS/aerie/blob/63bb374ca4346d9c5ed5390b3e8e8a8c30c5b2dc/db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java#L652

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Good catch - I can add an `assertFalse(res.next())` - or read the resultset into a list and check that its length is one

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Good catch, I'll update the comment :+1:

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java
Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java]: If I recall correctly, it was something to the effect of:
```
ERROR:  function allocate_dataset_partitions(bigint) does not exist
```

Comment (Mythicaeda) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/CreateDatasetPartitionsAction.java]: Makes sense. [The `apply` method in this class](https://github.com/NASA-AMMOS/aerie/pull/833/files#diff-86342dd34fb02577133fccc89429aee5b12a9c350b4e5a07a7d7f2c613a9e253L20-R23) takes (took?) a long, not an int.

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Done

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Comment updated

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java
Comment (mattdailis) [reply scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java]: I swapped out the `break` for a `return` here

File: scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/graphql/GraphQLParsers.java
Comment (mattdailis) [reply scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/graphql/GraphQLParsers.java:1]: I ended up reverting some of the parsing changes on the scheduler - so this pattern is now used again.

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:119]: Is the `+00` how you're forcing it to grab the results in UTC instead of in your local TZ?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java]: Why is this statement not needed in other tests?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:1]: Minor: There are some now-unused imports in this file

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java]: Minor: Import statement is now unused

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java]: Why do we need to call `microsOfPGInterval` here?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java]: Why do `TYPE_SCROLL_SENSITIVE` and `CONCUR_READ_ONLY` need to be specified here?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Minor: Unused Import

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: What exactly is `count(1)` selecting?

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java
Comment (Mythicaeda) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/PlanCollaborationTests.java]: Minor: This is the same check as in [L604-L613](https://github.com/NASA-AMMOS/aerie/pull/833/files#diff-269125a2369e062ea57b1a0ae448f5e159d3c8f69552c13ace30c65533e246b2R604-R613), so it should be written the same.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java]: Minor: Unused import. Additionally, `java.time.Instant` is also an unused import.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java]: Minor: The parameter `epoch` is now unused.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java
Comment (Mythicaeda) [inline merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java:111]: How does this work if this is called to parse, for example, the start offset of a span that starts two months into a year long simulation dataset? It looks like the method would throw an `IllegalArgumentException`, which I'd argue is unexpected behavior for calling [`GetSpanRecords.get()`](https://github.com/NASA-AMMOS/aerie/blob/prototype/use-org-postgresql-jdbc/merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/GetSpanRecords.java#L50) on a valid dataset.

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java
Comment (Mythicaeda) [inline scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java]: Why do we need to check `notifications.length`? Not only should it not be 0 based on the description of `getNotifciations` (which states that it will return either an array with at least one element or null), but if a 0 length array is returned, then won't it skip the foreach loop?

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java
Comment (Mythicaeda) [inline scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java:40]: This loop should be `while (!Thread.currentThread().isInterrupted())`, as otherwise we're waiting on trying to put something into the `notificationQueue` in order to detect an interrupted signal.

File: scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java
Comment (Mythicaeda) [inline scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/ListenSchedulerCapability.java]: The statement `Thread.currentThread().interrupt();` doesn't do anything here, since we return from the Thread's `run` method immediately after.

File: merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java
Comment (mattdailis) [reply merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java:44]: In the merlin worker, we could conceivably call it using a try-with-resources, since the lifetime of the data source is in lexical scope. In the servers, I'm less sure how to go about this - I would ask a similar question about `jetty`.

Something to ponder :+1:

Comment (Mythicaeda) [reply merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java:44]: Yeah. From what I recall, a shutdown hook will work for SIGINT (and [this SO discussion](https://stackoverflow.com/questions/2541475/capture-sigint-in-java) seems to support that), but that doesn't affect SIGKILL.

Either way, I'm resolving this conversation so that it doesn't block this PR.

File: merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java
Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java:111]: I did some testing on `develop`, and observed that `spans` never use the `months` field of intervals :+1: I manually updated the start offset of a span to include months, and constraints checking threw an exception:

Screenshot 2023-05-31 at 5 10 55 PM

I did the same with years - and found an interesting difference between the old and new jdbc drivers. On develop, a span with a start offset of `1 year` in the database is rewritten as `P12M` by the jdbc driver, and throws the same error as in the screenshot above. (Mystifyingly, it does this only for prepared statements - a regular `executeQuery` with the same SQL results in `P1Y`, as you would expect). The new driver behaves as you would expect.

I'll poke at it a bit more - but the impression I'm getting is that this PR changes the error messages, but does not cause previously error-free code paths to error.

I would like to take a holistic look at how we use intervals, and make sure we're being consistent (or at least the inconsistencies are documented) - that can be a follow-up to this PR

Comment (Mythicaeda) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java:111]: Awesome! Would it be possible to include this check as a test for the `microsOfPGInterval` function?

I agree that further investigation is a separate body of work, and just including that test would be sufficient to resolve the concern my initial comment brought up.

Comment (mattdailis) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java:111]: Could you help me understand which check you're referring to?

Comment (Mythicaeda) [reply merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresParsers.java:111]: The check where span records with a start offset greater than or equal to a month/year are produced and then fetched via `GetSpanRecords.get()`. (Or, equivalently, that `spans` never uses the 'month' field of intervals)

So something (really roughly) like:
```java
@Test
void microsFNTest(){
   final var planDuration = "2Years";
   createPlan(planDuration);
   for(int i = 0; i < 24; i++) 
        createActivity(i+" Months");
   createActivity("1 Year, 1 Month, 1 Day") // Feb 2nd effectively
   simulate(); // A simulation that will post to DB
   assertNotThrows(GetSpanRecords.get(simResults));
   // Further assertions could possibly be made about what the startOffsets looks like
   // but the important assertion is that fetching these results does not throw the IllegalArgumentException
}
```

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java]: I introduced it to try to get around this assertion error:
```
Expected :0 years 0 mons 0 days 0 hours 0 mins -3600.0 secs
Actual   :0 years 0 mons 0 days -1 hours 0 mins 0.0 secs
```

Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java]: If we hard-code the expected value, we can remove the need for `microsOfPGInterval` here:

```java
assertEquals(new PGInterval("-1 hours"), newOffsetFromPlanStart);
```

The whole test would then become:

```java
@Test
void shouldCalculatePlanDatasetOffsetOnPlanDatasetInsertWithNonNullDatasetId() throws SQLException {
  final var newPlanId = insertPlan(missionModelId, "2020-1-1 01:00:00");

  final var planDatasetInsertRes = connection.createStatement()
                                             .executeQuery(
                                                 """
                                                     INSERT INTO plan_dataset (plan_id, dataset_id)
                                                     VALUES (%s, %s)
                                                     RETURNING *;"""
                                                     .formatted(newPlanId, planDatasetRecord.dataset_id())
                                             );
  planDatasetInsertRes.next();
  final var newOffsetFromPlanStart = new PGInterval(planDatasetInsertRes.getString("offset_from_plan_start"));
  planDatasetInsertRes.close();

  assertEquals(new PGInterval("-1 hours"), newOffsetFromPlanStart);
}
```

A potential downside here is that this test would now assume that the previously inserted plan had a start time of `2020-1-1 00:00:00`. Perhaps a comment stating this assumption would be sufficient.

Comment (Mythicaeda) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/MerlinDatabaseTests.java]: ~~I don't think we need to state that in a comment, since it's part of the `insertPlan` statement (either implicitly if we use the `insertPlan(int)` or explicitly as part of the `final var newPlanId` line.~~

Sorry, misread what you were saying. I'm not against a comment, but I think it's still included if you look at the definition of `planDatasetRecord` (or repeat the plan and dataset insert of the `beforeEach` with locals at the top of this method)

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:119]: The `+00` makes sure that the created plan's start time is interpreted as UTC - without it, I get that the inserted plan's start time is `2020-1-1 00:00:00-08`, which equates to `2020-1-1 08:00:00+00`

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java]: Hmm, without it I get this test failure, but I'm having trouble rationalizing it at the moment:
```
Expected :2020-01-02 00:00:00+00
Actual   :2020-01-01 16:00:00-08
```

Those two times have more than a timezone difference 🤔 Perhaps the timezone setting has an effect on the logic in `get_approximate_start_time`

Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java]: This seems relevant, if a bit old: https://stackoverflow.com/a/30878862

Comment (Mythicaeda) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java]: No they don't. Midnight UTC is 16:00 the day prior ET (24-8=16)

Comment (Mythicaeda) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java]: Back to the original `connection.prepareStatement` line, it would make most sense for it to go in the `DatabaseTestHelper.startDatabase`, rather than trying to "catch" the side effects of not putting it there in various tests (ie here and in `MerlinDatabaseTests`).

Comment (mattdailis) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java]: > No they don't. Midnight UTC is 16:00 the day prior ET (24-8=16)

Whew! Thanks for checking me on that - that's a relief

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java
Comment (Mythicaeda) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/AnchorTests.java:119]: Is this still needed with declaring the TZ to UTC?

**Review by Mythicaeda — APPROVED** [review]
Look's great! Let's get this thing 🚢

**Pull request merged**
